standalone/include/psen_scan_v2_standalone/data_conversion_layer/io_pin_data.h

```diff
@@ -42,31 +39,115 @@ namespace monitoring_frame
  */
 namespace io
 {
-inline static uint32_t createID(size_t byte_n, size_t bit_n)
+static constexpr uint32_t NUMBER_OF_INPUT_BYTES{ RAW_CHUNK_LOGICAL_INPUT_SIGNALS_IN_BYTES };
+static constexpr uint32_t NUMBER_OF_OUTPUT_BYTES{ RAW_CHUNK_OUTPUT_SIGNALS_IN_BYTES };
+
+//! @throws std::out_of_range if byte_location >= NUMBER_OF_INPUT_BYTES or bit_location >= 8
+static inline LogicalInputType getInputType(std::size_t byte_location, std::size_t bit_location)
+{
+  return LOGICAL_INPUT_BITS.at(byte_location).at(bit_location);
+}
+
+//! @throws std::out_of_range if byte_location >= NUMBER_OF_INPUT_BYTES or bit_location >= 8
+static inline std::string getInputName(std::size_t byte_location, std::size_t bit_location)
 {
-  return byte_n * 8 + bit_n;
+  return LOGICAL_INPUT_BIT_TO_NAME.at(getInputType(byte_location, bit_location));
+}
+
+//! @throws std::out_of_range if byte_location >= NUMBER_OF_OUTPUT_BYTES or bit_location >= 8
+static inline OutputType getOutputType(std::size_t byte_location, std::size_t bit_location)
+{
+  return OUTPUT_BITS.at(byte_location).at(bit_location);
+}
+
+//! @throws std::out_of_range if byte_location >= NUMBER_OF_OUTPUT_BYTES or bit_location >= 8
+static inline std::string getOutputName(std::size_t byte_location, std::size_t bit_location)
+{
+  return OUTPUT_BIT_TO_NAME.at(getOutputType(byte_location, bit_location));
 }
 
 /**
  * @brief Represents the IO PIN field of a monitoring frame.
  */
-struct PinData
+class PinData
 {
-  using States = std::vector<PinState>;
-  States input{};
-  States output{};
+public:
+  //! @throws std::out_of_range if byte_location >= NUMBER_OF_INPUT_BYTES or bit_location >= 8
+  bool inputPinState(std::size_t byte_location, std::size_t bit_location) const;
+  //! @throws std::out_of_range if byte_location >= NUMBER_OF_OUTPUT_BYTES or bit_location >= 8
+  bool outputPinState(std::size_t byte_location, std::size_t bit_location) const;
+
+  //! @throws std::out_of_range if byte_location >= NUMBER_OF_INPUT_BYTES or bit_location >= 8
+  void inputPinState(std::size_t byte_location, std::size_t bit_location, bool pin_state);
+  //! @throws std::out_of_range if byte_location >= NUMBER_OF_OUTPUT_BYTES or bit_location >= 8
+  void outputPinState(std::size_t byte_location, std::size_t bit_location, bool pin_state);
+
+  std::array<std::bitset<8>, NUMBER_OF_INPUT_BYTES>& inputState();
+  std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES>& outputState();
+  const std::array<std::bitset<8>, NUMBER_OF_INPUT_BYTES>& inputState() const;
+  const std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES>& outputState() const;
+
+  bool operator==(const PinData& pin_data) const;
+
+private:
+  std::array<std::bitset<8>, NUMBER_OF_INPUT_BYTES> input_states_{};
+  std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES> output_states_{};
 };
```

Comment:
We have now two ways of accessing the arrays with the same effect:
```
bool old_state = pin_data.inputPinState(4, 3);
pin_data.inputPinState(4, 3, true);
```
```
bool old_state = pin_data.inputState().at(4).test(3);
pin_data.inputState().at(4).set(3);
```
Do you think we have increased readability through the methods `inputPinState(...)`? If not, we could as well go back to the struct:
```
struct PinData
{
std::array<std::bitset<8>, NUMBER_OF_INPUT_BYTES> input_states{};
std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES> output_states{};
};
```

Reply:
I think I like it better as a struct. The convenience methods are rarely used (in production code) anyways. The "set"-methods never and the "get"-methods only once.